### ISSUE TITLE
Agent connections cause spam in Caddy's log file

### ISSUE THREAD

**Issue by whalehub** [Ylianst/MeshCentral#431]
I'm using the reverse proxy Caddy (v1.0.1) in front of MeshCentral (v0.3.9-u) and today I noticed the following error messages popping up a few times per second, every second in Caddy's log file:

```
2019/08/14 16:39:34 [ERROR] failed to copy buffer:  read tcp 127.0.0.1:38078->127.0.0.1:7450: use of closed network connection
2019/08/14 16:39:34 [ERROR] failed to copy buffer:  read tcp 127.0.0.1:38080->127.0.0.1:7450: use of closed network connection
2019/08/14 16:39:34 [ERROR] failed to copy buffer:  read tcp 127.0.0.1:38082->127.0.0.1:7450: use of closed network connection
2019/08/14 16:39:34 [ERROR] failed to copy buffer:  read tcp 127.0.0.1:38086->127.0.0.1:7450: use of closed network connection
2019/08/14 16:39:34 [ERROR] failed to copy buffer:  read tcp 127.0.0.1:38088->127.0.0.1:7450: use of closed network connection
2019/08/14 16:39:34 [ERROR] failed to copy buffer:  read tcp 127.0.0.1:38090->127.0.0.1:7450: use of closed network connection
2019/08/14 16:39:35 [ERROR] failed to copy buffer:  read tcp 127.0.0.1:38094->127.0.0.1:7450: use of closed network connection
2019/08/14 16:39:35 [ERROR] failed to copy buffer:  read tcp 127.0.0.1:38096->127.0.0.1:7450: use of closed network connection
2019/08/14 16:39:35 [ERROR] failed to copy buffer:  read tcp 127.0.0.1:38100->127.0.0.1:7450: use of closed network connection
```

I managed to find out that this issue was introduced in v0.3.9-r by downgrading several times.

**Comment by Ylianst** [comment]
Hi. Yes, v0.3.9-r is a bad version, upgrade to -s or better and this should be fixed. Let me know if that works.

**Comment by whalehub** [comment]
@Ylianst I was already using the latest version (v0.3.9-u) when I first noticed this issue, I only downgraded all the way to v0.3.9-r to find out in which version the issue was introduced.

**Comment by Ylianst** [comment]
So, is port 7450 the MeshCentral main port? (normally 443)
What version of nodejs are you using ("node -v" to find out).

v0.3.9-r did introduce the new 404 error page. If the is the problem, you can check by going to "node_modules/meshcentral/webserver.js" on a later server and searching and deleting the following 5 lines in that file. Then, restart the server.

```
            // Handle 404 error
            obj.app.use(function (req, res, next) {
                var domain = getDomain(req);
                res.status(404).render(obj.path.join(obj.parent.webViewsPath, isMobileBrowser(req) ? 'error404-mobile' : 'error404'), { title: domain.title, title2: domain.title2 });
            })
```

Probably good to backup your existing "node_modules" folder, "npm install meshcentral" see if that fixed it, if not, revert back to the old "node_modules".

**Comment by whalehub** [comment]
@Ylianst I'm using 7450 as the alias port for port 443 and I'm running NodeJS v11.15.0 on Debian 10.

I just deleted those lines and restarted the server and it fixed the issue. Is it possible to rewrite that piece of code so as to avoid causing this issue?

**Comment by Ylianst** [comment]
Oh nice! I am very happy we are on to something. Is it possible to send over a anonymized version of your Caddy's configuration file? I may need to install it myself to figure out exactly what it going on. I was wondering if the health check would be causing the issue, but the events in the log are very quick. I will probably need to look at the network traffic to see the problem. Sending a nicer 404 error message should not cause this.

FYI: I am on vacation until next Wednesday, but will be around here to solve issues and answer questions until then and Bryan will be working (solving the PowerShell issue). So, probably best to keep these lines removed for now. Worst case, I can add a config.json option to remove the nice 404 page when I get back, but would rather get to the bottom of this.

**Comment by whalehub** [comment]
@Ylianst Here's my Caddy config for MeshCentral:

```
meshcentral.example.com {

  proxy / 127.0.0.1:7450 {
    transparent
    websocket
  }

  gzip {
    ext *
  }

  header / {
    Strict-Transport-Security "max-age=31536000; includeSubDomains; preload"
    X-XSS-Protection "1; mode=block"
    X-Content-Type-Options "nosniff"
    X-Frame-Options "allow-from https://example.com/"
    Referrer-Policy "same-origin"
    Expect-CT "max-age=86400, enforce, report-uri=\"https://example.report-uri.com/r/d/ct/enforce\""
    Feature-Policy "camera 'none'; geolocation 'none'; microphone 'none'"
    Content-Security-Policy "
      default-src 'none';
      base-uri 'self';
      script-src 'self';
      style-src 'self';
      img-src *;
      font-src 'self' data:;
      connect-src 'self' data:;
      media-src 'self' data:;
      object-src 'none';
      worker-src 'self';
      manifest-src 'self';
      frame-ancestors 'self' https://example.com/;
      form-action 'self';
      upgrade-insecure-requests;
      block-all-mixed-content;
      report-uri https://example.report-uri.com/r/d/csp/enforce;"
    -Server
    -X-Powered-By
  }

  expires {
    match .css$ 7d
    match .gif$ 7d
    match .ico$ 7d
    match .jpg$ 7d
    match .js$ 7d
    match .json$ 7d
    match .pdf$ 7d
    match .png$ 7d
    match .svg$ 7d
  }

  tls {
    protocols tls1.2 tls1.3
    dns gandiv5
    ciphers ECDHE-ECDSA-AES128-GCM-SHA256 ECDHE-ECDSA-WITH-CHACHA20-POLY1305 ECDHE-ECDSA-AES256-GCM-SHA384
    key_type p384
    must_staple
    wildcard
  }
}
```

I'm using NodeJS v11.15.0.

**Comment by Ylianst** [comment]
Published MeshCentral v0.4.0-f with a fix for the 404 page. If you had sub-domains in MeshCentral, the 404 page would break them. It's fixed now.

In the unlikely chance the 404 error page is still a problem, you can put "nice404" : false in the settings section of the config.json to remove it. However, I am hoping this will not be needed.

```
{
  "settings": {
    "Cert": "sample.com",
    "Nice404": false
  }
}
```

**Comment by whalehub** [comment]
@Ylianst I just upgraded to v0.4.0-f and can confirm that the issue is fixed now. Thanks!
